subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,608 @@
+		ctx->fds[0].fd = ctx->client_ctx.transport.tcp.sock;
+	}
+
+	ctx->fds[0].events = ZSOCK_POLLIN;
+	ctx->nfds = 1;
+}
+
+static void clear_fds(struct shell_mqtt *ctx)
+{
+	ctx->nfds = 0;
+}
+
+static int wait(int timeout)
+{
+	int ret = 0;
+
+	if (sh_mqtt->nfds > 0) {
+		ret = zsock_poll(sh_mqtt->fds, sh_mqtt->nfds, timeout);
+		if (ret < 0) {
+			LOG_ERR("poll error: %d", errno);
+		}
+	}
+
+	return ret;
+}
+
+static void mqtt_connect_work(struct k_work *work);
+
+/*
+ * Work routine to process MQTT packet and keep alive MQTT connection
+ */
+static void mqtt_process_work(struct k_work *work)
+{
+	int64_t remaining = APP_SLEEP_MSECS;
+	int64_t start_time = k_uptime_get();
+	struct shell_mqtt *ctx = NULL;
+	int rc;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	/* Listen to the port for 500 ms as defined by APP_SLEEP_MSECS */
+	while (remaining > 0) {
+		if (wait(remaining)) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+			}
+		}
+
+		rc = mqtt_live(&ctx->client_ctx);
+		if (rc != 0 && rc != -EAGAIN) {
+			LOG_ERR("mqtt_live error: %d", rc);
+			return;
+		} else if (rc == -ENOTCONN) {
+			(void)mqtt_disconnect(&ctx->client_ctx);
+			k_work_init_delayable(&ctx->mqtt_work, mqtt_connect_work);
+			(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+			return;
+		} else if (rc == 0) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+			}
+		}
+
+		remaining = APP_SLEEP_MSECS + start_time - k_uptime_get();
+	}
+
+	/* Reschedule the work */
+	k_work_init_delayable(&ctx->mqtt_work, mqtt_process_work);
+	(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+}
+
+static int mqtt_subscribe_config(struct mqtt_client *const client)
+{
+	/* subscribe to config information */
+	struct mqtt_topic subs_topic = { .topic = { .utf8 = sh_mqtt->sub_topic,
+						    .size = strlen(sh_mqtt->sub_topic) },
+					 .qos = MQTT_QOS_1_AT_LEAST_ONCE };
+	const struct mqtt_subscription_list subs_list = { .list = &subs_topic,
+							  .list_count = 1U,
+							  .message_id = 1U };
+	int ret;
+
+	ret = mqtt_subscribe(client, &subs_list);
+	if (ret) {
+		LOG_ERR("Failed to subscribe to %s item, error %d", subs_topic.topic.utf8, ret);
+	}
+
+	return ret;
+}
+
+static void mqtt_subscribe_work(struct k_work *work)
+{
+	struct shell_mqtt *ctx = NULL;
+	int ret;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	ret = mqtt_subscribe_config(&ctx->client_ctx);
+	if (ret == 0) {
+		/* Change the work to process/keepalive */
+		LOG_DBG("Scheduling MQTT process & keepalive work");
+		k_work_init_delayable(&ctx->mqtt_work, mqtt_process_work);
+		(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+	} else {
+		k_work_init_delayable(&ctx->mqtt_work, mqtt_subscribe_work);
+		(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
```

Comment:
I've modified it to `mqtt_abort` and reschedule the `mqtt_connect_work` once the `mqtt_subscribe_config` fails.

Reply:
@rlubos does this apply to other mqtt APIs as well (i.e. `mqtt_publish`, `mqtt_connect`), as in the application has to `mqtt_abort` and start over from `mqtt_connect` again?

Reply:
Theoretically, when anything goes wrong with the underlying stream and error is reported, the mqtt library should do the celanup on it's own (`mqtt_abort()` is nothing more than closing a socket and clearing the internal state). However, as for now, there's no harm in doing this from the application as well, it might protect the app in case there's some undiscovered corner case where this is not done. And it is more future proof, as this might change in the future, there've been some discussion already, that mqtt lib shall not close the socket on it's own, but rather only shut down the connection, and let the application disconnect/abort to avoid rare race condition at the socket level.

Reply:
@rlubos thanks for the insights